### PR TITLE
Feature/50 file history

### PR DESCRIPTION
**Pull request by mbehlendorf**

### CODE REVIEW

File: src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGitCommitRepository.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package com.cloudogu.smeagol.wiki.infrastructure;
+
+import com.cloudogu.smeagol.wiki.domain.*;
+import com.google.common.base.Throwables;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.eclipse.jgit.lib.PersonIdent;
```
Comment (cesmarvin) [inline src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGitCommitRepository.java]: MINOR Remove this unused import 'org.eclipse.jgit.lib.PersonIdent'. [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGitCommitRepository.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import org.eclipse.jgit.revwalk.RevCommit;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.io.File;
```
Comment (cesmarvin) [inline src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGitCommitRepository.java]: MINOR Remove this unused import 'java.io.File'. [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGitCommitRepository.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import java.io.IOException;
+import java.time.Instant;
```
Comment (cesmarvin) [inline src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGitCommitRepository.java]: MINOR Remove this unused import 'java.time.Instant'. [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGitPageRepository.java [hunk 4]
```diff
@@ -15,6 +15,8 @@
 import java.time.Instant;
```
Comment (cesmarvin) [inline src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGitPageRepository.java]: MINOR Remove this unused import 'java.time.Instant'. [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/cloudogu/smeagol/wiki/domain/CommitRepository.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package com.cloudogu.smeagol.wiki.domain;
+
+public interface CommitRepository {
```
Comment (cesmarvin) [inline src/main/java/com/cloudogu/smeagol/wiki/domain/CommitRepository.java]: MAJOR Annotate the "CommitRepository" interface with the @FunctionalInterface annotation [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AS1609)

File: src/main/java/com/cloudogu/smeagol/wiki/infrastructure/PageController.java [hunk 6]
```diff
@@ -122,6 +141,10 @@ private Path getMoveTo() {
             }
             return Path.valueOf(moveTo);
         }
+        private CommitId getRestore() {
+            if (Strings.isNullOrEmpty(restore)) return null;
```
Comment (cesmarvin) [inline src/main/java/com/cloudogu/smeagol/wiki/infrastructure/PageController.java]: MINOR At most one statement is allowed per line, but 2 statements were found on this line. [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AS00122)

File: src/main/java/com/cloudogu/smeagol/wiki/usecase/Commits.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package com.cloudogu.smeagol.wiki.usecase;
+
+import com.cloudogu.smeagol.Account;
+import com.cloudogu.smeagol.AccountService;
+import com.cloudogu.smeagol.wiki.domain.*;
+
+import java.time.Instant;
+
+public class Commits {
```
Comment (cesmarvin) [inline src/main/java/com/cloudogu/smeagol/wiki/usecase/Commits.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AS1118)

File: src/main/java/com/cloudogu/smeagol/wiki/infrastructure/HistoryResourceAssembler.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+import static org.springframework.hateoas.mvc.ControllerLinkBuilder.linkTo;
+
+public class HistoryResourceAssembler extends ResourceAssemblerSupport<History, HistoryResource> {
+
+    public HistoryResourceAssembler() {
+        super(HistoryController.class, HistoryResource.class);
+    }
+
+
+    private final CommitResourceAssembler assembler = new CommitResourceAssembler();
```
Comment (cesmarvin) [inline src/main/java/com/cloudogu/smeagol/wiki/infrastructure/HistoryResourceAssembler.java]: MINOR Move this variable to comply with Java Code Conventions. [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AS1213)

File: src/main/java/com/cloudogu/smeagol/wiki/infrastructure/HistoryResourceAssembler.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Override
+    public HistoryResource toResource(History history) {
+        List<CommitResource> commitsAsResources = new ArrayList<CommitResource>(history.getCommits().size());
```
Comment (cesmarvin) [inline src/main/java/com/cloudogu/smeagol/wiki/infrastructure/HistoryResourceAssembler.java]: MAJOR Replace the type specification in this constructor call with the diamond operator ("<>"). [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AS2293)

File: src/main/java/com/cloudogu/smeagol/wiki/infrastructure/GitClient.java [hunk 10]
```diff
@@ -123,6 +126,21 @@ public File file(String path) {
...
+    public List<RevCommit> findCommits(String path) throws IOException, GitAPIException {
+        Git git = open();
+        List<RevCommit> commits = new ArrayList<>();
+        Iterator<RevCommit> iterator = git.log()
+                .addPath(path)
+                .call()
+                .iterator();
+
+        if (iterator.hasNext()) {
+            iterator.forEachRemaining(revCommit -> commits.add(revCommit));
```
Comment (cesmarvin) [inline src/main/java/com/cloudogu/smeagol/wiki/infrastructure/GitClient.java]: MINOR Replace this lambda with a method reference. [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AS1612)

File: src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGit.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+import com.cloudogu.smeagol.wiki.domain.*;
+import org.eclipse.jgit.lib.PersonIdent;
+import org.eclipse.jgit.revwalk.RevCommit;
+
+import java.time.Instant;
+
+/**
+ * Util class for the ScmGit repositories.
+ */
+public final class ScmGit {
```
Comment (cesmarvin) [inline src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGit.java]: MAJOR Add a private constructor to hide the implicit public one. [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AS1118)

**Comment by cesmarvin** [comment]
SonarQube analysis reported 14 issues
* ![MAJOR][MAJOR] 5 major
* ![MINOR][MINOR] 8 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ScmGitPageRepository.java#L8](https://github.com/cloudogu/smeagol/blob/9710f183c863e7d84449a410606a5340c98772a2/src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGitPageRepository.java#L8): Remove this unused import 'org.eclipse.jgit.lib.PersonIdent'. [rule](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3AUselessImportCheck)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGit.java (hunk 11 above)
Comment (mbehlendorf) [reply src/main/java/com/cloudogu/smeagol/wiki/infrastructure/ScmGit.java]: done

**Comment by cesmarvin** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Pull request merged**